File: android/app/src/main/java/com/alpaquitay/contractpay/network/ApiService.kt
package com.alpaquitay.contractpay.network

import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Response
import retrofit2.http.*

/**
 * Retrofit API interface defining all backend endpoints.
 */
interface ApiService {

    // ──────────────────────────────────────────
    // Authentication
    // ──────────────────────────────────────────

    @POST("auth/token")
    suspend fun getToken(
        @Body body: Map<String, String>
    ): Response<TokenResponse>

    // ──────────────────────────────────────────
    // Contracts
    // ──────────────────────────────────────────

    /**
     * Upload a contract file (PDF or TXT).
     * Uses multipart form data for file upload.
     */
    @Multipart
    @POST("contracts/upload")
    suspend fun uploadContract(
        @Part contract: MultipartBody.Part,
        @Part("metadata") metadata: RequestBody? = null
    ): Response<UploadResponse>

    /**
     * Upload a text contract directly (no file).
     */
    @POST("contracts/upload-text")
    suspend fun uploadTextContract(
        @Body body: Map<String, String>
    ): Response<UploadResponse>

    /**
     * Get contract details by ID.
     */
    @GET("contracts/{id}")
    suspend fun getContract(
        @Path("id") contractId: String
    ): Response<ContractDetailResponse>

    /**
     * Verify contract integrity (recompute hash).
     */
    @POST("contracts/{id}/verify")
    suspend fun verifyContract(
        @Path("id") contractId: String
    ): Response<VerifyResponse>

    /**
     * List all contracts for the current user.
     */
    @GET("contracts")
    suspend fun listContracts(
        @Query("page") page: Int = 1,
        @Query("limit") limit: Int = 20
    ): Response<ContractListResponse>

    // ──────────────────────────────────────────
    // Payments
    // ──────────────────────────────────────────

    /**
     * Create a Clover order for payment.
     */
    @POST("payments/clover/create-order")
    suspend fun createCloverOrder(
        @Body body: Map<String, Any>
    ): Response<CloverOrderResponse>

    /**
     * Verify Clover payment completion.
     */
    @POST("payments/clover/verify")
    suspend fun verifyCloverPayment(
        @Body body: Map<String, String>
    ): Response<PaymentVerifyResponse>

    /**
     * Get payment status for a contract.
     */
    @GET("payments/status/{contractId}")
    suspend fun getPaymentStatus(
        @Path("contractId") contractId: String
    ): Response<PaymentStatusResponse>
}

// ──────────────────────────────────────────────
// Response Data Classes
// ──────────────────────────────────────────────

data class TokenResponse(
    val token: String,
    val type: String,
    val expiresIn: String,
    val userId: String
)

data class UploadResponse(
    val message: String,
    val contract: ContractData
)

data class ContractData(
    val id: String,
    val originalName: String,
    val mimeType: String? = null,
    val fileSize: Long? = null,
    val sha256Hash: String,
    val hashGeneratedAt: String,
    val status: String,
    val createdAt: String? = null
)

data class ContractDetailResponse(
    val contract: ContractDetail
)

data class ContractDetail(
    val id: String,
    val originalName: String,
    val mimeType: String,
    val fileSize: Long,
    val sha256Hash: String,
    val hashGeneratedAt: String,
    val status: String,
    val paymentStatus: String,
    val paymentProvider: String,
    val createdAt: String,
    val updatedAt: String
)

data class VerifyResponse(
    val contractId: String,
    val integrity: IntegrityResult
)

data class IntegrityResult(
    val isValid: Boolean,
    val storedHash: String,
    val computedHash: String,
    val verifiedAt: String,
    val status: String
)

data class ContractListResponse(
    val contracts: List<ContractData>,
    val pagination: PaginationData
)

data class PaginationData(
    val page: Int,
    val limit: Int,
    val total: Int,
    val pages: Int
)

data class CloverOrderResponse(
    val message: String,
    val order: CloverOrder
)

data class CloverOrder(
    val orderId: String,
    val merchantId: String,
    val amountCents: Int,
    val contractId: String,
    val status: String
)

data class PaymentVerifyResponse(
    val message: String,
    val verification: Map<String, Any>,
    val contractStatus: String?
)

data class PaymentStatusResponse(
    val contractId: String,
    val paymentStatus: String,
    val paymentProvider: String,
    val paymentId: String?,
    val amountCents: Int,
    val contractStatus: String
)

File: frontend/src/apiService.ts
/**
 * apiService.ts — Handles all communication with the Alpaquitay backend.
 * Manages JWT token and provides typed methods for each endpoint.
 */

// ─── Types ───

export interface ContractData {
  _id: string;
  id?: string;
  originalName: string;
  mimeType?: string;
  fileSize?: number;
  sha256Hash: string;
  hashGeneratedAt: string;
  status: 'uploaded' | 'verified' | 'paid' | 'tampered' | 'expired';
  paymentStatus?: string;
  paymentProvider?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface IntegrityResult {
  isValid: boolean;
  storedHash: string;
  computedHash: string;
  verifiedAt: string;
  status: string;
}

export interface PaymentSession {
  transactionId: string;
  checkoutUrl: string;
  contractId: string;
  amountCents: number;
  currency: string;
  status: string;
  expiresAt: string;
}

// ─── Token Management ───

let authToken: string | null = null;

function setToken(token: string) {
  authToken = token;
  localStorage.setItem('alpacachain_token', token);
}

function getToken(): string | null {
  if (!authToken) {
    authToken = localStorage.getItem('alpacachain_token');
  }
  return authToken;
}

export function clearToken() {
  authToken = null;
  localStorage.removeItem('alpacachain_token');
}

export function isAuthenticated(): boolean {
  return !!getToken();
}

// ─── Base Fetch ───

async function apiFetch(path: string, options: RequestInit = {}): Promise<any> {
  const token = getToken();
  const headers: Record<string, string> = {
    ...(options.headers as Record<string, string> || {}),
  };

  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  // Don't set Content-Type for FormData (browser handles it with boundary)
  if (!(options.body instanceof FormData)) {
    headers['Content-Type'] = 'application/json';
  }

  const res = await fetch(`/api${path}`, {
    ...options,
    headers,
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({ error: `HTTP ${res.status}` }));
    throw new Error(err.error || err.message || `Request failed: ${res.status}`);
  }

  return res.json();
}

// ─── Auth ───

export async function login(userId: string, apiKey: string): Promise<string> {
  const data = await apiFetch('/auth/token', {
    method: 'POST',
    body: JSON.stringify({ userId, apiKey }),
  });
  setToken(data.token);
  return data.token;
}

// ─── Contracts ───

export async function uploadContractFile(file: File, metadata?: Record<string, string>): Promise<ContractData> {
  const formData = new FormData();
  formData.append('contract', file);
  if (metadata) {
    formData.append('metadata', JSON.stringify(metadata));
  }

  const data = await apiFetch('/contracts/upload', {
    method: 'POST',
    body: formData,
  });
  return data.contract;
}

export async function uploadContractText(content: string, title: string): Promise<ContractData> {
  const data = await apiFetch('/contracts/upload-text', {
    method: 'POST',
    body: JSON.stringify({ content, title }),
  });
  return data.contract;
}

export async function getContract(id: string): Promise<ContractData> {
  const data = await apiFetch(`/contracts/${id}`);
  return data.contract;
}

export async function listContracts(page = 1, limit = 50): Promise<{ contracts: ContractData[]; total: number }> {
  const data = await apiFetch(`/contracts?page=${page}&limit=${limit}`);
  return { contracts: data.contracts, total: data.pagination.total };
}

export async function verifyContract(id: string): Promise<IntegrityResult> {
  const data = await apiFetch(`/contracts/${id}/verify`, { method: 'POST' });
  return data.integrity;
}

// ─── Payments ───

export async function createFiservSession(params: {
  contractId: string;
  amountCents: number;
  customerEmail: string;
  currency?: string;
}): Promise<PaymentSession> {
  const data = await apiFetch('/payments/fiserv/create-session', {
    method: 'POST',
    body: JSON.stringify({
      contractId: params.contractId,
      amountCents: params.amountCents,
      customerEmail: params.customerEmail,
      currency: params.currency || 'USD',
      returnUrl: window.location.origin,
    }),
  });
  return data.session;
}

export async function getPaymentStatus(contractId: string) {
  return apiFetch(`/payments/status/${contractId}`);
}
